lib/datadog/security/reactive/engine.rb

```diff
@@ -0,0 +1,41 @@
+require 'datadog/security/reactive/address_hash'
+require 'datadog/security/reactive/subscriber'
+
+module Datadog
+  module Security
+    module Reactive
+      class Engine
+        def initialize
+          @data = {}
+          @subscribers = AddressHash.new { |h, k| h[k] = [] }
+          @children = []
+        end
+
+        def subscribe(*addresses, &block)
+          @subscribers[addresses.freeze] << Subscriber.new(&block).freeze
+        end
+
+        def publish(address, data)
+          # check if someone has address subscribed
+          if @subscribers.addresses.include?(address)
+
+            # someone will be interested, set data
+            @data[address] = data
+
+            # find candidates i.e address groups that contain the just posted address
+            @subscribers.with(address).each do |addresses|
+              # find targets to the address group containing the posted address
+              subscribers = @subscribers[addresses]
+
+              # is all data for the targets available?
+              if (addresses - @data.keys).empty?
+                h = addresses.each_with_object({}) { |a, h| h[a] = @data[a] }
+                subscribers.each { |s| s.call(*h.values) }
+              end
+            end
+          end
+        end
```

Comment:
This looks like something that will happen quite often; we may want to keep an eye out in terms of performance, as I think we can do better here if needed (but only if we measure this to indeed have a noticeable impact).